[Mikhil Anand] iP

Sasmik23 opened the pull request:
## DukePro
> "Your mind is for having ideas, not holding them"

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
And it is **FREE!**

Features:

- [ ] Managing tasks
- [ ] Managing deadlines 
- [ ]  Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Sasmik23 committed:
Level0

Sasmik23 committed:
Level1

Sasmik23 committed:
Level2 - add,list

Sasmik23 committed:
Level-3: Mark as Done

Sasmik23 committed:
Level 4 - todos, events, deadlines

Sasmik23 committed:
A-TextUITesting

Sasmik23 committed:
level-5 error handling

Sasmik23 committed:
level-6 delete

Sasmik23 committed:
added comments and javadocs

Sasmik23 committed:
level-7: save

Sasmik23 committed:
level 8: date

Sasmik23 committed:
A-MoreOOP

Sasmik23 committed:
A-Packages

Sasmik23 committed:
Merge branch 'add-gradle-support' of https://github.com/Sasmik23/ip into gradle

* 'add-gradle-support' of https://github.com/Sasmik23/ip:
  Bump gradle and lib version
  Add Gradle support

Sasmik23 committed:
A-packages update and A-gradle update

Sasmik23 committed:
A-graddle run added

Sasmik23 committed:
A-Junit todoTest

Sasmik23 committed:
A-CodingStandard

Sasmik23 committed:
A-JavaDoc

Sasmik23 committed:
level 9 find

Sasmik23 committed:
Merge branch 'branch-A-JavaDoc'

Sasmik23 committed:
resolved merge conflicts

Sasmik23 committed:
A-Jar update'

wxwern commented on src/main/java/duke/filehandler/Storage.java:
```diff
@@ -0,0 +1,122 @@
+            return date;
+        } catch (Exception e) {
+            System.out.println(e);
+           throw new DukeException("Cannot read time from file");
```
There appears to be an indentation error here:
```suggestion
            throw new DukeException("Cannot read time from file");
```

wxwern commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,63 @@
+        } else if(action.equals("unmark")){
+            this.isDone = false;
+            System.out.println("OK, I've marked this task as not done yet:");
+        }
```
I like that you prefixed the boolean variable with `is`, so it sounds like a boolean!

Similarly, please add whitespace around Java keywords etc., as per code standards.

```suggestion
        if (action.equals("mark")) {
            this.isDone = true;
            System.out.println("Nice! I've marked this task as done:");
        } else if (action.equals("unmark")) {
            this.isDone = false;
            System.out.println("OK, I've marked this task as not done yet:");
        }
```

wxwern commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,63 @@
+     * reduce size of tasks and print remaining number
+     */
+    public void delete(){
+        size = size -1;
```
```suggestion
        size = size - 1;
```

wxwern commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,63 @@
+    }
+    @Override
+    public String toString() {
+        return "[" + this.getStatusIcon()+ "] " + this.description;
```
```suggestion
        return "[" + this.getStatusIcon() + "] " + this.description;
```

wxwern commented on src/main/java/duke/parsers/InputParser.java:
```diff
@@ -0,0 +1,205 @@
+     * @param task type of task - todo,event,deadline,mark/unmark
+     * @throws DukeException
+     */
+    public static void inputChecker(String[] str, String task) throws DukeException {
```
As per our standards, JavaDoc descriptions should [follow the form provided here](https://se-education.org/guides/conventions/java/intermediate.html#comments). Notably, it should only have `/**` at the start, have a line break between description and parameters, and should have punctuation.

```suggestion
    /**
     * Checks user inputs, if invalid throws DukeException.
     *
     * @param str User input split by each word.
     * @param task Type of task, may be one of: "todo", "event", "deadline", "mark/unmark".
     * @throws DukeException If the input is invalid.
     */
```

I notice this also applies to other JavaDocs as well, so do update them appropriately.

wxwern commented on src/main/java/duke/parsers/InputParser.java:
```diff
@@ -0,0 +1,205 @@
+            return false;
+        }
+        //list out each task from Duke.tasks ArrayList
+        else if (input.equals("list")) {
```
For `if` blocks that you want to add comments on, perhaps consider moving the comment to within the code block? That way, your code follows Egyptian-style brackets.

```suggestion
        } else if (input.equals("list")) {
            //list out each task from Duke.tasks ArrayList
```

This applies to other similar sections as well.

Additionally, with this many cases, perhaps you might want to consider using `switch`-`case`? IDEs are likely to also be able to auto-convert them for you.

wxwern commented on src/main/java/duke/parsers/InputParser.java:
```diff
@@ -0,0 +1,205 @@
+                int index = Integer.parseInt(splitStr[1]);
+                Task item = tasks.get(index - 1);
+                item.setAction(splitStr[0]);
+                if (toStore){  Storage.saveTasks(tasks); };
```
Conditionals should be put in a separate line as per coding standards.
```suggestion
                if (toStore) {
                    Storage.saveTasks(tasks); 
                };
```

wxwern commented on src/main/java/duke/filehandler/Storage.java:
```diff
@@ -0,0 +1,122 @@
+            Scanner reader = new Scanner(taskFile);
+            while (reader.hasNextLine()) {
+                String data = reader.nextLine();
+                if (data.length() > 0){
+                    if (data.charAt(1)=='T'){
+                        Todo todo = new Todo (data.substring(7));
+                        if (data.charAt(4) == 'X'){
+                            todo.setAction("mark");
+                        }
+                        tasks.add(todo);
+                    } else if (data.charAt(1)=='D'){
+                        String dates = data.split(":",2)[1];
+                        String desc = data.split("\\(")[0];
+                        Date date = fileDateParser(dates.substring(1,dates.length()-1));
+                        Deadline deadline = new Deadline(desc.substring(7), date);
+                        if (data.charAt(4) == 'X'){
+                            deadline.setAction("mark");
+                        }
+                        tasks.add(deadline);
+                    } else if (data.charAt(1) == 'E'){
+                        String[] dates = data.split(":");
+                        String desc = data.split("\\(")[0];
+                        Date from = fileDateParser(dates[1].substring(1,dates[1].length()-2));
+                        Date to = fileDateParser(dates[2].substring(1,dates[2].length()-1));
+                        Event event = new Event(desc.substring(7), from,to);
+                        if (data.charAt(4) == 'X'){
+                            event.setAction("mark");
+                        }
+                        tasks.add(event);
+                    }
+                }
+            }
+        } catch( IOException e){
```
Please add or remove appropriate spacing here, primarily after commas, and around Java keywords and operators, as per our coding standards.

For example, in this segment:
```suggestion
                if (data.length() > 0) {
                    if (data.charAt(1) == 'T') {
                        Todo todo = new Todo(data.substring(7));
                        if (data.charAt(4) == 'X') {
                            todo.setAction("mark");
                        }
                        tasks.add(todo);
                    } else if (data.charAt(1) == 'D') {
                        String dates = data.split(":", 2)[1];
                        String desc = data.split("\\(")[0];
                        Date date = fileDateParser(dates.substring(1, dates.length() - 1));
                        Deadline deadline = new Deadline(desc.substring(7), date);
                        if (data.charAt(4) == 'X') {
                            deadline.setAction("mark");
                        }
                        tasks.add(deadline);
                    } else if (data.charAt(1) == 'E') {
                        String[] dates = data.split(":");
                        String desc = data.split("\\(")[0];
                        Date from = fileDateParser(dates[1].substring(1, dates[1].length() - 2));
                        Date to = fileDateParser(dates[2].substring(1, dates[2].length() - 1));
                        Event event = new Event(desc.substring(7), from, to);
                        if (data.charAt(4) == 'X') {
                            event.setAction("mark");
                        }
                        tasks.add(event);
                    }
                }
            }
        } catch (IOException e) {
```

I also noticed that these formatting issues show up in other areas as well, so do update them appropriately. To save time, you might want to use `codestyle` to automatically detect them, and configure your IDE to automatically format them as well.

wxwern commented on src/main/java/duke/filehandler/Storage.java:
```diff
@@ -0,0 +1,122 @@
+     * @return tasks as an ArrayList
+     * @throws DukeException
+     */
+
```
Documentation should not be separated from the method by a newline.
```suggestion
```

wxwern commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# Duke project template
+# duke.duke project template
```
Was this accidentally refactored? Seems like a lot of references to "Duke" got erroneously replaced with "duke.duke".

wxwern reviewed (commented):
Overall, most of the code is alright, though quite a few parts have coding standard violations, primarily on spacing issues. Do fix them as needed!

flexibo commented on src/main/java/duke/filehandler/Storage.java:
```diff
@@ -0,0 +1,122 @@
+                return tasks;
+            }
+            Scanner reader = new Scanner(taskFile);
+            while (reader.hasNextLine()) {
+                String data = reader.nextLine();
+                if (data.length() > 0){
```
Will it be better to avoid triple nesting if statements? You can make use of continue statements to move onto the next loop, so for example

```java
            while (reader.hasNextLine()) {
                String data = reader.nextLine();
                if (data.length() <= 0){
                    continue}
                }
                // Rest of code...
```
Reference links from the CS2103 code quality guide:
[Avoid Deep Nesting (Basic)](https://nus-cs2103-ay2324s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting)
[Make the Happy Path Prominent (Advanced)](https://nus-cs2103-ay2324s1.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent)

flexibo commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,63 @@
+    protected String description;
+    protected boolean isDone;
+
+    //total number of tasks
```
nice job adding comments to variables that may not be obvious at first glance appropriately 😄

flexibo commented on src/main/java/duke/parsers/InputParser.java:
```diff
@@ -0,0 +1,205 @@
+                System.out.println(e.getMessage());
+            }
+        }
+        //create event object, splitting the due date by "/" and stripping off the to: and from:
```
nice comment explaining each if block parts accordingly, code also seems to be in a logical order, easy for me to understand, and also following the KISS (Keep It Simple, Stupid) convention.

flexibo commented on src/main/java/duke/tasks/Todo.java:
```diff
@@ -0,0 +1,17 @@
+
+    public Todo(String description) {
+        super(description);
+        System.out.println("Got it. I've added this task:");
```
Not that it is wrong but, is it possible to separate the print statements with the Task Classes themselves? So instead of printing here, you print in the caller function instead. It is best practice to separate the display and the data if possible. 

so for example 

Todo.java
```java 
public Todo(String description) {
    super(description);
}
```
Duke.java / wherever you create an object of this class
```
list.add(new Todo("blablabla");
System.out.println("Got it. I've added this task:");
System.out.println("  " + this);
System.out.println("Now you have " + super.size + " tasks in the list.");
```

flexibo commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,30 @@
+
+    protected Date from;
+    protected Date to;
+    protected String fromFormat;
+    protected String toFormat;
```
Good job separating the names between Date and String and having similar naming formats, might be better to change the name of the variable here as it does not seem to be clear as to what format this variable is.
```suggestion
    protected String fromStringFormat;
    protected String toStringFormat;
```

flexibo reviewed (commented):
Overall nice and simple code, just added a few suggestions for improvement in code style.

Sasmik23 committed:
testing javafx

Sasmik23 committed:
Level-10 GUI

Sasmik23 committed:
Improve code quality

To help with readability

Adhere to code quality guidelines (ie. spacing)

Standardise with everyone else

Sasmik23 committed:
Merge pull request #2 from Sasmik23/branch-A-CodeQuality

Improve code quality

Sasmik23 committed:
C-Tagging

Sasmik23 committed:
Update UI

Added welcome and ending messages, added padding to text and changed window sizing
Added new StorageException Class

Sasmik23 committed:
updated readme, yet to add screenshot

Sasmik23 committed:
Added screenshot of UI and fixed bug in displaying tags

Sasmik23 committed:
update readme

Sasmik23 committed:
change jar file name

Sasmik23 committed:
fix bug in displaying size

Sasmik23 committed:
- saving tasks creates new directory
- fixed bug in todo file

Sasmik23 committed:
- stores data only when user exits programme
- added 2 more junit tests

Sasmik23 committed:
changed pictures to comply with copyright issues

Sasmik23 committed:
update Ui.png

Sasmik23 committed:
Refactor code based on peer reviews

Abstract commands from inputParser, add header comments for classes
methods and add a few more tests.
